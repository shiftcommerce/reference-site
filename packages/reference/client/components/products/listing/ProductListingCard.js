// Libraries
import React from 'react'
import classNames from 'classnames'
import Link from 'next/link'

// Components
import ProductPrice from '../display/ProductPrice'

// Objects
import Image from '../../../objects/Image'

class ProductListingCard extends React.Component {
  renderRatingStars (product) {
    let rating = (product.rating || 0)

    return (
      [1, 2, 3, 4, 5].map((key, index) =>
        <span key={key} className={classNames({ 'c-product-listing-card__rating--fill': key <= rating, 'c-product-listing-card__rating--empty': key > rating })}>
          &#9733;
        </span>
      )
    )
  }

  render () {
    let {
      product
    } = this.props

    const assetFile = product.asset_files[0]

    return (
      <div className='c-product-listing-card'>
        <div className='c-product-listing-card__body'>
          <div className='c-product-listing-card__gallery  c-image'>
            <Link href={`/slug?slug=${product.canonical_path}`} as={product.canonical_path}>
              <Image className='c-product-listing-card__image'
                src={assetFile && assetFile.s3_url}
                alt={(assetFile && assetFile.alt_text) || product.title}
                aria-label={product.title} />
            </Link>
          </div>
          <div className='c-product-listing-card__title' >
            <p role='heading' aria-level='1'>
              {product.title}
            </p>
          </div>
          <div className='c-product-listing-card__price'>
            <ProductPrice variants={product.variants} />
          </div>
          <div className='c-product-listing-card__rating'>
            { this.renderRatingStars(product) }
          </div>
        </div>
      </div>
    )
  }
}

export default ProductListingCard
