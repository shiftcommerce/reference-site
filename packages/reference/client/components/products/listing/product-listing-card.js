// Libraries
import React from 'react'
import classNames from 'classnames'
import Link from 'next/link'

// Components
import ProductPrice from '../display/product-price'

// Objects
import { Rating, LazyLoad } from 'shift-react-components'

class ProductListingCard extends React.Component {
  renderImageAndTitle () {
    const {
      assetFileAltText,
      assetFileUrl,
      title,
      imageHeight,
      imageWidth
    } = this.props

    const DEFAULT_IMAGE_DIMENSION = 280

    // Needed so that the aspect ratio can be calculated
    // if it is undefined then lazyload div wont display correctly.
    // TODO: remove this when we specify height/width in Algolia.
    const plpImageHeight = imageHeight || DEFAULT_IMAGE_DIMENSION
    const plpImageWidth = imageWidth || DEFAULT_IMAGE_DIMENSION

    return (
      <>
        <div className='c-product-listing-card__gallery c-image'>
          <LazyLoad className='c-product-listing-card__image u-image-shadow'
            src={assetFileUrl}
            alt={assetFileAltText || title}
            aria-label={title}
            imageHeight={plpImageHeight}
            imageWidth={plpImageWidth}
          />
        </div>
        <div className='c-product-listing-card__title' >
          <p role='heading' aria-level='1'>
            { title }
          </p>
        </div>
      </>
    )
  }

  render () {
    const {
      className,
      maxPrice,
      minPrice,
      productPath,
      productRating
    } = this.props

    return (
      <div className={classNames('c-product-listing-card', className)}>
        <div className='c-product-listing-card__body'>
          <Link href={`/slug?slug=${productPath}`} as={productPath}>
            <a>
              { this.renderImageAndTitle() }
            </a>
          </Link>
          <div className='c-product-listing-card__price'>
            <ProductPrice minPrice={minPrice} maxPrice={maxPrice} />
          </div>
          <div className='c-product-listing-card__rating'>
            <Rating rating={productRating} />
          </div>
        </div>
      </div>
    )
  }
}

export default ProductListingCard
