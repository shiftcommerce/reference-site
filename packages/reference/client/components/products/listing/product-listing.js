// Libraries
import { Component } from 'react'
import classNames from 'classnames'

// Objects
import Breadcrumb from '../../../objects/breadcrumb'

// Components
import { AlgoliaHits } from '../../search/algolia/algolia-hits'
import AlgoliaFilters from '../../search/algolia/algolia-filters'

import ProductMenu from './product-menu'
import ProductMenuOptions from './product-menu-options'

export class ProductListing extends Component {
  constructor (props) {
    super(props)
    this.state = {
      showFilters: false
    }
  }

  toggleFiltering = () => {
    this.setState({
      showFilters: !this.state.showFilters
    })
  }

  render () {
    return (
      <>
        <ProductMenu title={this.props.title} body={'Optional text prop not yet setup but available'} />
        <Breadcrumb />
        <div className='c-product-listing-wrapper'>
          <AlgoliaFilters showFilters={this.state.showFilters} toggleFiltering={this.toggleFiltering} />
          <div className={classNames('c-product-listing')}>
            <div className='c-product-listing__menu'>
              <ProductMenuOptions showFilters={this.state.showFilters} toggleFiltering={this.toggleFiltering} />
            </div>
            <AlgoliaHits />
          </div>
        </div>
      </>
    )
  }
}

export default ProductListing
