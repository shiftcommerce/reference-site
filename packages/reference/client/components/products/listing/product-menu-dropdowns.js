const ProductMenuDropdowns = () => (
  <div className='c-product-listing__menu-dropdowns'>
    <div className='c-product-listing__menu-dropdown c-product-listing__menu-dropdown--category'>
      <select className='c-product-listing__menu-dropdown-select'><option>category</option></select>
    </div>
    <div className='c-product-listing__menu-dropdown c-product-listing__menu-dropdown--colour'>
      <select className='c-product-listing__menu-dropdown-select'><option>colour</option></select>
    </div>
    <div className='c-product-listing__menu-dropdown c-product-listing__menu-dropdown--size'>
      <select className='c-product-listing__menu-dropdown-select'><option>size</option></select>
    </div>
    <div className='c-product-listing__menu-dropdown c-product-listing__menu-dropdown--price'>
      <select className='c-product-listing__menu-dropdown-select'><option>price</option></select>
    </div>
    <div className='c-product-listing__menu-dropdown c-product-listing__menu-dropdown--sortby'>
      <select className='c-product-listing__menu-dropdown-select'><option>sort by</option></select>
    </div>
  </div>
)

export default ProductMenuDropdowns
