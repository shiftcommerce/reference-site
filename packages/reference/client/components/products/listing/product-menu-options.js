// Libraries
import { PureComponent } from 'react'
import { SortBy } from 'react-instantsearch-dom'
import getConfig from 'next/config'

// Components
import AlgoliaRefinements from '../../search/algolia/algolia-refinements'

// Objects
import Button from '../../../objects/button'

const {
  publicRuntimeConfig: {
    ALGOLIA_INDEX_NAME
  }
} = getConfig()

class ProductMenuOptions extends PureComponent {
  render () {
    const { toggleFiltering } = this.props

    return (
      <div className='c-product-listing__menu-options'>
        <div className='c-product-listing__menu-options-filters'>
          <h2 className='c-product-listing__menu-options-filters-title'>Filters</h2>
          <div className='c-product-listing__menu-options-filters-applied'>
            <AlgoliaRefinements />
          </div>
          <Button className='c-product-listing__menu-options-filters-button' onClick={toggleFiltering} />
        </div>
        <div className='c-product-listing__menu-options-sort-by'>
          <h2 className='c-product-listing__menu-options-sort-by-title'>Sort by:</h2>
          <SortBy defaultRefinement={ALGOLIA_INDEX_NAME}
            items={[
              { value: ALGOLIA_INDEX_NAME, label: 'Featured' },
              { value: `${ALGOLIA_INDEX_NAME}_product_rating_asc`, label: 'Rating asc.' },
              { value: `${ALGOLIA_INDEX_NAME}_product_rating_desc`, label: 'Rating desc.' }
            ]}
          />
          <Button className='c-product-listing__menu-options-sort-by-button u-hidden-d' />
        </div>
      </div>
    )
  }
}

export default ProductMenuOptions
