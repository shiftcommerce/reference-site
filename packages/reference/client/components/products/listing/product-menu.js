import { connectStateResults } from 'react-instantsearch-dom'

export const ProductMenu = (props) => {
  const searchQuery = props.searchState.query
  let searchTitle = 'Showing all products'

  if (searchQuery) {
    searchTitle = `Search "${searchQuery}"`
  }

  return (
    <div className='c-product-listing__menu-description'>
      <h1 className='c-product-listing__menu-description-title'>
        { props.title || searchTitle }
      </h1>
    </div>
  )
}

export default connectStateResults(ProductMenu)
