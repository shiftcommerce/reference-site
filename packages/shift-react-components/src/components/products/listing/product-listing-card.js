// Libraries
import React, { Component } from 'react'
import classNames from 'classnames'

// Objects
import LazyLoad from '../../../objects/lazy-load'
import Link from '../../../objects/link'
import Rating from '../../../objects/rating'

// Components
import ProductPrice from '../display/product-price'

class ProductListingCard extends Component {
  renderImageAndTitle () {
    const {
      assetFileAltText,
      assetFileUrl,
      title,
      imageHeight,
      imageWidth
    } = this.props

    const DEFAULT_IMAGE_DIMENSION = 280

    // Needed so that the aspect ratio can be calculated
    // if it is undefined then lazyload div wont display correctly.
    // TODO: remove this when we specify height/width in Algolia.
    const plpImageHeight = imageHeight || DEFAULT_IMAGE_DIMENSION
    const plpImageWidth = imageWidth || DEFAULT_IMAGE_DIMENSION

    return (
      <>
        <div className='c-product-listing-card__gallery o-image'>
          <LazyLoad className='c-product-listing-card__image u-image-shadow'
            src={assetFileUrl}
            alt={assetFileAltText || title}
            aria-label={title}
            imageHeight={plpImageHeight}
            imageWidth={plpImageWidth}
          />
        </div>
        <div className='c-product-listing-card__title' >
          <p role='heading' aria-level='1'>
            { title }
          </p>
        </div>
      </>
    )
  }

  render () {
    const {
      className,
      maxPrice,
      minPrice,
      productPath,
      productRating
    } = this.props

    return (
      <div className={classNames('c-product-listing-card', className)}>
        <div className='c-product-listing-card__body'>
          <Link href={`/slug?slug=${productPath}`} as={productPath}>
            { this.renderImageAndTitle() }
          </Link>
          <div className='c-product-listing-card__price'>
            <ProductPrice minPrice={minPrice} maxPrice={maxPrice} />
          </div>
          <div className='c-product-listing-card__rating'>
            <Rating rating={productRating} />
          </div>
        </div>
      </div>
    )
  }
}

export default ProductListingCard
