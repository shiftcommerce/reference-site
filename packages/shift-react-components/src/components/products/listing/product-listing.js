// Libraries
import React, { Component } from 'react'
import classNames from 'classnames'

// Components
import ProductMenu from '../listing/product-menu'
import ProductMenuOptions from '../listing/product-menu-options'
import SearchFilters from '../../search/search-filters'
import SearchHits from '../../search/search-hits'

// Objects
import { Breadcrumb } from '../../../objects/breadcrumb'

export class ProductListing extends Component {
  constructor (props) {
    super(props)
    this.state = {
      filtersShown: false
    }
  }

  toggleFiltering = () => {
    if (this.state.filtersShown) {
      document.body.classList.remove('modal-open')
    } else {
      document.body.classList.add('modal-open')
    }

    this.setState({
      filtersShown: !this.state.filtersShown
    })
  }

  render () {
    const { title, facets, indexName, indexNameWithoutDefaultSortOrder } = this.props

    return (
      <>
        <ProductMenu title={title} />
        <Breadcrumb />
        <div className='c-product-listing-wrapper'>
          <SearchFilters facets={facets} filtersShown={this.state.filtersShown} toggleFiltering={this.toggleFiltering} />
          <div className={classNames('c-product-listing')}>
            <div className='c-product-listing__menu'>
              <ProductMenuOptions
                indexName={indexName}
                indexNameWithoutDefaultSortOrder={indexNameWithoutDefaultSortOrder}
                toggleFiltering={this.toggleFiltering}
              />
            </div>
            <SearchHits />
          </div>
        </div>
      </>
    )
  }
}

export default ProductListing
