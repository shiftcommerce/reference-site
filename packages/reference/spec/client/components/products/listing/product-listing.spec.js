// Component
import { ProductListing } from '../../../../../client/components/products/listing/product-listing'
import AlgoliaFilters from '../../../../../client/components/search/algolia/algolia-filters'

import ProductMenu from '../../../../../client/components/products/listing/product-menu'
import ProductMenuOptions from '../../../../../client/components/products/listing/product-menu-options'

// Fixtures
import searchResultState from '../../../../fixtures/search-result-state'

// Mock out the next/config
jest.mock('next/config', () => () => ({
  publicRuntimeConfig: {
    ALGOLIA_API_KEY: 'FAKE-API-KEY',
    ALGOLIA_APP_ID: 'FAKE-APP-ID',
    ALGOLIA_INDEX_NAME: 'FAKE-INDEX'
  }
}))

test('renders correctly', () => {
  // Arrange
  const searchState = {
    query: 'tray'
  }

  const onSearchStateChange = (nextSearchState) => (nextSearchState)

  // Act
  const wrapper = shallow(
    <ProductListing
      resultsState={searchResultState}
      onSearchStateChange={onSearchStateChange}
      searchState={searchState}
    />
  )

  const rootTree = wrapper.find('c-product-listing')

  // Assert
  expect(wrapper).toMatchSnapshot()
  expect(rootTree.find(ProductMenu))
  expect(rootTree.find(ProductMenuOptions))
  expect(rootTree.find(AlgoliaFilters))
})
