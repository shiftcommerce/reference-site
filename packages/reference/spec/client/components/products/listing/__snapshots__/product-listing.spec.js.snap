// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<React.Fragment>
  <ProductMenu
    body="Optional text prop not yet setup but available"
  />
  <Breadcrumb />
  <div
    className="c-product-listing-wrapper"
  >
    <AlgoliaFilters
      showFilters={false}
      toggleFiltering={[Function]}
    />
    <div
      className="c-product-listing"
    >
      <div
        className="c-product-listing__menu"
      >
        <ProductMenuOptions
          showFilters={false}
          toggleFiltering={[Function]}
        />
      </div>
      <AlgoliaInfiniteHits(AlgoliaResults) />
    </div>
  </div>
</React.Fragment>
`;
